File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def minDepth(root: TreeNode): Int = {
        if(root == null) return 0
        helper(List(root))
    }
    
    def helper(list: List[TreeNode]): Int = {
        if(list.isEmpty) return 0
        if(list.exists(o => o.left == null && o.right == null)) 1
        else helper(list.flatMap(o => List(o.left, o.right).filterNot(_ == null))) + 1
    }
}

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.rs
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun minDepth(root: TreeNode?): Int = with(ArrayDeque<TreeNode>()) {
      root?.let { add(it) }
      generateSequence(1) { (it + 1).takeIf { isNotEmpty() } }
        .firstOrNull {
          (1..size).any {
            with(poll()) {
              left?.let { add(it) }
              right?.let { add(it) }
              left == null && right == null
            }
          }
        } ?: 0
    }
}